astropy/modeling/tests/test_quantities_evaluation.py

```diff
@@ -0,0 +1,538 @@
+    """
+
+    class TestModelC(Fittable1DModel):
+        a = Parameter()
+        input_units = lambda a: a.unit * u.m
+        @staticmethod
+        def evaluate(x, a):
+            return a * x * u.s ** 2
+
+    m = TestModelC(a=1 / u.s)
+    assert m(3 * u.m / u.s) == 3 * u.m
+
+    with pytest.raises(UnitsError) as exc:
+        m(1 * u.m)
+    assert exc.value.args[0] == ("Units of input 'x', m (length), could "
+                                 "not be converted to required input units of "
+                                 "m / s (speed)")
+
+
+def test_input_units_fixed_unit():
+    """
+    Test setting input_units to a fixed unit
+    """
+
+    class TestModelD(Fittable1DModel):
+        input_units = u.m
+        @staticmethod
+        def evaluate(x):
+            return 2 * x
+
+    m = TestModelD()
+
+    with pytest.raises(UnitsError) as exc:
+        m(1 * u.s)
+    assert exc.value.args[0] == ("Units of input 'x', s (time), could "
+                                 "not be converted to required input units of "
+                                 "m (length)")
+
+    assert m(2 * u.m) == 4 * u.m
+    assert_quantity_allclose(m(300 * u.cm), 6 * u.m)
+
+
+def test_input_units_instance_specific():
+    """
+    Test changing input_units on-the-fly for a specific instance
+    """
+
+    class TestModelD(Fittable1DModel):
+        input_units = u.m
+        @staticmethod
+        def evaluate(x):
+            return 2 * x
+
+    m1 = TestModelD()
+    m2 = TestModelD()
+    m2.input_units = u.s
+
+    assert m1(1 * u.m) == 2 * u.m
+    with pytest.raises(UnitsError) as exc:
+        m2(1 * u.m)
+    assert exc.value.args[0] == ("Units of input 'x', m (length), could not be "
+                                 "converted to required input units of s (time)")
+    assert m2(1 * u.s) == 2 * u.s
+
+
+# We now test that we are able to set equivalencies for input/output units. We
+# don't need to repeat all the above tests, we simply choose the case where the
+# input/output units are fixed. Equivalencies are set by the
+# ``input_equivalencies`` and ``output_equivalencies`` attributes for models.
+
+
+@pytest.mark.xfail
+def test_input_units_equivalencies():
+    """
+    Test that input_equivalencies is recognized
+    """
+
+    class TestModel(Fittable1DModel):
+        input_units = u.m
+        input_equivalencies = u.spectral()
+        @staticmethod
+        def evaluate(x):
+            return 2 * x
+
+    m = TestModel()
+    assert_quantity_allclose(m(3 * u.nm), 6 * u.nm)
+
+    # We now check it works with frequency. Note that we should make sure
+    # we evaluate the model outside the context manager to make sure that we
+    # are picking up the equivalency from input_equivalencies. The context
+    # manager is only there because assert_quantity_allclose doesn't support
+    # equivalencies.
+    result = m(400 * u.GHz)
+    with u.set_enabled_equivalencies(u.spectral()):
```

Comment:
If you are paranoid about the context manager you could test `np.allclose(m(400 * u.GHz).value, 800.)`.